### ISSUE TITLE
Support initialization of / assignment to tuples via iterable expressions?

### ISSUE THREAD

**Issue by bradcray** [chapel-lang/chapel#18969]
Since tuples support iteration:

```chapel
for e in (1, 2, 3) do ...
```

I often find myself thinking that I should be able to initialize a tuple or assign to a tuple using an iterable expression as well, for example:

```chapel
// or any of these without the `var` or with a singleton tuple variable `my3Tuple`
var (i1, i2, i3) = [i in 1..3] i;
var (a1, a2, a3) = A[1..3];
var (s1, s2, s3) = "I'm a tuple".split(" ");
```

since it feels like a form of zippered iteration.  Should we support such forms?

**Comment by mppf** [comment]
It's interesting because in all of your examples the RHS number of elements could vary at runtime, while the LHS the number of elements is known at compile time. So I would suppose this pattern would have to do bounds checking or something, but maybe that's not a big deal.

**Comment by bradcray** [comment]
I agree that a runtime bounds check would be needed to head off errors.

**Comment by aconsroe-hpe** [comment]
Would the last example support assigning to both an array and a tuple? Which would be the default?

```chapel
var parts = "I'm a tuple".split(" "); // today this is an [{0..2}] string
var parts: 3 * string = "I'm a tuple".split(" "); // maybe specifying the type would change what happens?

proc overload(x: [{0..2}] string) {}
proc overload(x: 3 * string) {}
overload("I'm a tuple".split(" "); // which one do I call?
```

We do already do runtime checking for something like:

```chapel
var parts1: [{0..3}] string = "I'm a tuple".split(" ");
/* Fails at runtime with error: zippered iterations have non-equal lengths */
```

I also use this in eg Python code so would be in support of having it. It does occur to me that all these examples are for a homogenous tuple (all same type) since the `iter` can only be of one type.

**Comment by vasslitvinov** [comment]
+1 to initializing a tuple from an iterable expression. Here are a couple of extensions:

* `[for param i in 1..N] f(i)` could be an N-tuple literal for a param N

* `var (a1, a2, ...) = an iterable expression;` or `var (a1, a2, _) = ...` (or some other syntax) would accept iterables producing >2 elements wtihout error; the iteration would (or not?) quit (as if with a `break;`) after producing the first 2 elements

**Comment by bradcray** [comment]
> Would the last example support assigning to both an array and a tuple?

Yes.

> Which would be the default?
> overload("I'm a tuple".split(" "); // which one do I call?

By this issue's proposal, I actually don't _think_ either would be the default because for them to work, we'd also need to support coercions from iterable expressions to arrays and/or tuples.  If you put `in` intents on both of them and we took this proposal, then I think it would become an ambiguity and you'd need to remove one of the overloads or cast to specify which to call (or maybe name the formals something different to distinguish and call-by-keyword to indicate which you wanted?).

> /* Fails at runtime with error: zippered iterations have non-equal lengths */
Oh, that's interesting.  We're often bad at length-checking for forall loops, which is what I was imagining was being used here, so that's encouraging.

**Comment by bradcray** [comment]
> [for param i in 1..N] f(i) could be an N-tuple literal for a param N

For my tastes, this would be too subtle.  I'd still expect an array with domain [1..N] rather than a tuple (whose indices are `0..<N`, note).

> var (a1, a2, ...) = an iterable expression; or var (a1, a2, _) = ... (or some other syntax) would accept iterables producing >2 elements wtihout error; the iteration would (or not?) quit (as if with a break;) after producing the first 2 elements

I find this attractive, but consider it orthogonal to this proposal in the sense that one could similarly want either without the other.  For example:

```chapel
var (a1, a2, ...) = myFiveTuple;
```
